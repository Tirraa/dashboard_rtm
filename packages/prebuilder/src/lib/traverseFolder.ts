import type { Stats } from 'fs';

import type { Filename, File, Path } from '../types/Metadatas';
import type { Thunks, Thunk } from '../types/LazyEvaluation';

type ArborescenceFragmentPromise = Promise<File[] | File>;

// https://github.com/vitest-dev/vitest/discussions/2484
const path = require('path');
const fs = require('fs/promises');

async function makeThunks(currentFolder: Path, __currentDeepPath: Path = currentFolder): Promise<Thunks<ArborescenceFragmentPromise>> {
  const files: Filename[] = await fs.readdir(currentFolder);
  const thunks: Thunks<ArborescenceFragmentPromise> = files.map((currentFilename: Filename) => {
    const maybeFilepath: Path = path.join(currentFolder, currentFilename);
    const thunk: Thunk<ArborescenceFragmentPromise> = async () => {
      const filepathStats: Stats = await fs.stat(maybeFilepath);
      return filepathStats.isDirectory()
        ? execute(makeThunks(maybeFilepath, path.join(__currentDeepPath, currentFilename)))
        : ({ directory: __currentDeepPath, name: currentFilename } satisfies File);
    };
    return thunk;
  });

  return Promise.all(thunks);
}

const execute = async (thunks: Promise<Thunks<ArborescenceFragmentPromise>>): Promise<File[]> =>
  (await Promise.all((await thunks).map((thunk) => thunk()))).flat();

const traverseFolder = async (rootFolder: Path): Promise<File[]> => execute(makeThunks(rootFolder));

export default traverseFolder;
