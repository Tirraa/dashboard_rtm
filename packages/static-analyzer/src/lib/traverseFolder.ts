import type { Stats } from 'fs';
import { readdirSync, statSync } from 'fs';
import { basename, extname, join } from 'path';
import type { File, Filename, Path } from '../types/metadatas';

const CURRENT_DIRECTORY_PREFIX = './';

export function traverseFolder(rootFolder: Path): File[] {
  const filesCollection: File[] = [];

  function traverse(currentFolder: Path, currentDeepPath: Path = currentFolder): void {
    const currentFolderFiles: Filename[] = readdirSync(currentFolder);

    for (const currentFilename of currentFolderFiles) {
      const filepath: Path = join(currentFolder, currentFilename);
      const stat: Stats = statSync(filepath);

      if (stat.isDirectory()) {
        traverse(filepath, join(currentDeepPath, currentFilename));
        continue;
      }

      const filename = basename(currentFilename, extname(currentFilename));
      const fileDirectory =
        !currentDeepPath.startsWith(CURRENT_DIRECTORY_PREFIX) && rootFolder.startsWith(CURRENT_DIRECTORY_PREFIX)
          ? CURRENT_DIRECTORY_PREFIX + currentDeepPath
          : currentDeepPath;

      filesCollection.push({ fileDirectory, filename });
    }
  }

  traverse(rootFolder);
  return filesCollection;
}

export default traverseFolder;
