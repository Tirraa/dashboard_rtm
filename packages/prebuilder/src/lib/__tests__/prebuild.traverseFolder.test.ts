// eslint-disable-next-line import/no-extraneous-dependencies
import { describe, expect, it } from 'vitest';

import traverseFolder from '../traverseFolder';

const INVALID_PATH = './bless/the/draco/these/rounds/holy/sorry/bout/your/luck';

describe('traverseFolder', () => {
  it('should return an arborescence, given a valid RootFolder (full)', () => {
    expect(traverseFolder('./packages/prebuilder/src/lib/__tests__/fakeDirectory')).toMatchSnapshot();
  });

  it('should return an arborescence, given a valid RootFolder (partial)', () => {
    expect(traverseFolder('./packages/prebuilder/src/lib/__tests__/fakeDirectory/bar')).toMatchSnapshot();
  });

  it('should return an arborescence, given a valid RootFolder (partial 2)', () => {
    expect(traverseFolder('./packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz/foo')).toMatchSnapshot();
  });

  it('should throw ENOENT, given invalid path', () => {
    try {
      traverseFolder(INVALID_PATH);
    } catch (e) {
      const interceptedError = e as Error;
      if ('code' in interceptedError) {
        expect(interceptedError.code).toBe('ENOENT');
      } else {
        throw new Error('Error code not found');
      }
    }
  });
});
