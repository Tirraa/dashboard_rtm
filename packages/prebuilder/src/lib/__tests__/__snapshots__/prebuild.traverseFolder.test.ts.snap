// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`traverseFolder > should return an arborescence, given a valid RootFolder (full) 1`] = `
[
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "filename": "file1",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "filename": "file2",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "filename": "file3",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz/foo",
    "filename": "file1",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar",
    "filename": "file1",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar",
    "filename": "file2",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz/bar",
    "filename": "file1",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz/bar/foo",
    "filename": "file1",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz/bar/foo",
    "filename": "file2",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz",
    "filename": "file1",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz",
    "filename": "file2",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz",
    "filename": "file3",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo/baz/bar",
    "filename": "file1",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo/baz/bar",
    "filename": "file2",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo/baz",
    "filename": "file1",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo",
    "filename": "file1",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo",
    "filename": "file2",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo",
    "filename": "file3",
  },
]
`;

exports[`traverseFolder > should return an arborescence, given a valid RootFolder (partial 2) 1`] = `
[
  {
    "fileDirectory": "./packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz/foo",
    "filename": "file1",
  },
]
`;

exports[`traverseFolder > should return an arborescence, given a valid RootFolder (partial) 1`] = `
[
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "filename": "file1",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "filename": "file2",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "filename": "file3",
  },
  {
    "fileDirectory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz/foo",
    "filename": "file1",
  },
  {
    "fileDirectory": "./packages/prebuilder/src/lib/__tests__/fakeDirectory/bar",
    "filename": "file1",
  },
  {
    "fileDirectory": "./packages/prebuilder/src/lib/__tests__/fakeDirectory/bar",
    "filename": "file2",
  },
]
`;
